pkg/extension/extension.go

```diff
@@ -0,0 +1,74 @@
+package extension
+
+import (
+	"context"
+
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/release"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/rest"
+)
+
+// ReconcilerExtension defines an extension for the reconciler.
+// It consists of several sub-interfaces.
+type ReconcilerExtension interface {
+	Name() string
+	BeginReconciliationExtension
+	EndReconciliationExtension
+}
+
+// BeginReconciliationExtension defines the extension point to execute at the beginning of a reconciliation flow.
+type BeginReconciliationExtension interface {
+	BeginReconcile(ctx context.Context, reconciliationContext *Context, obj *unstructured.Unstructured) error
+}
+
+// EndReconciliationExtension defiens the extension point to execute at the end of a reconciliation flow.
+type EndReconciliationExtension interface {
+	EndReconcile(ctx context.Context, reconciliationContext *Context, obj *unstructured.Unstructured) error
+}
+
+// NoOpReconcilerExtension implements all extension methods as no-ops and can be used for convenience
+// when only a subset of the available methods needs to be implemented.
+type NoOpReconcilerExtension struct{}
```

Comment:
Do we need `NoOpReconcilerExtension`? Instead can we can let the users handle this, since as we keep on adding extension points, we would have to keep updating this. In an ideal scenario, we would not want breaking changes when we add extension points, but in case we still end up having predefined interfaces, would this be necessary?